lib/maths/CBoostedTreeLeafNodeStatistics.cc

```diff
@@ -434,24 +532,72 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
                     regularization.treeSizePenaltyMultiplier() -
                     regularization.depthPenaltyMultiplier() *
                         (2.0 * penaltyForDepthPlusOne - penaltyForDepth)};
-
         SSplitStatistics candidate{gain,
                                    h.trace() / static_cast<double>(m_NumberLossParameters),
                                    feature,
                                    splitAt,
                                    std::min(leftChildRowCount, c - leftChildRowCount),
                                    2 * leftChildRowCount < c,
-                                   assignMissingToLeft};
+                                   assignMissingToLeft,
+                                   leftChildRowCount,
+                                   c - leftChildRowCount};
         LOG_TRACE(<< "candidate split: " << candidate.print());
 
         if (candidate > result) {
             result = candidate;
+            childrenGainStatsGlobal = childrenGainStatsPerFeature;
         }
     }
+    if (m_Derivatives.numberLossParameters() > 2) {
+        // short-circuit the bound computation for the multi-class case
+        result.s_LeftChildMaxGain = INF;
+        result.s_RightChildMaxGain = INF;
+    } else if (result.s_Gain > 0) {
+        double childPenaltyForDepth{regularization.penaltyForDepth(m_Depth + 1)};
+        double childPenaltyForDepthPlusOne{regularization.penaltyForDepth(m_Depth + 2)};
+        double childPenalty{regularization.treeSizePenaltyMultiplier() +
+                            regularization.depthPenaltyMultiplier() *
+                                (2.0 * childPenaltyForDepthPlusOne - childPenaltyForDepth)};
+        result.s_LeftChildMaxGain =
+            0.5 * childMaxGain(childrenGainStatsGlobal.s_GLeft,
+                               childrenGainStatsGlobal.s_MinLossLeft, lambda) -
+            childPenalty;
+
+        result.s_RightChildMaxGain =
+            0.5 * childMaxGain(childrenGainStatsGlobal.s_GRight,
+                               childrenGainStatsGlobal.s_MinLossRight, lambda) -
+            childPenalty;
+    }
 
     LOG_TRACE(<< "best split: " << result.print());
 
     return result;
 }
+
+double CBoostedTreeLeafNodeStatistics::childMaxGain(double gChild,
+                                                    double minLossChild,
+                                                    double lambda) const {
```

Comment:
I think it might be worthwhile writing a brief outline of what this is doing, i.e. something along the lines of: "This computes the maximum possible gain we can expect splitting a child node given we know the sum of the positive (g^+) and negative gradients (g^-) at its parent, the minimum curvature on the positive and negative gradient set (hmin^+ and hmin^-) and largest and smallest gradient (gmax and gmin, respectively). The highest possible gain consistent with these constraints can be shown to be: (g^+)^2 / (hmin^+ g^+ / gmax + lambda) + (g^-)^2 / (hmin^- g^- / gmin + lambda)".

Reply:
I added the explanation comment as you suggested.